pkg/controller/daemon/daemoncontroller_test.go

```diff
@@ -708,6 +717,63 @@ func TestDaemonKillFailedPods(t *testing.T) {
 	}
 }
 
+// DaemonSet should not launch a pod on a tainted node when the pod doesn't tolerate that taint.
+func TestTaintedNodeDaemonDoesNotLaunchUntoleratePod(t *testing.T) {
+	manager, podControl, _ := newTestController()
+
+	node := newNode("tainted", nil)
+	setNodeTaint(node)
+	manager.nodeStore.Add(node)
+
+	ds := newDaemonSet("untolerate")
+	manager.dsStore.Add(ds)
+
+	syncAndValidateDaemonSets(t, manager, ds, podControl, 0, 0)
+}
+
+// DaemonSet should launch a pod on a tainted node when the pod can tolerate that taint.
+func TestTaintedNodeDaemonLaunchesToleratePod(t *testing.T) {
+	manager, podControl, _ := newTestController()
+
+	node := newNode("tainted", nil)
+	setNodeTaint(node)
+	manager.nodeStore.Add(node)
+
+	ds := newDaemonSet("tolerate")
+	setDaemonSetTeleration(ds)
+	manager.dsStore.Add(ds)
+
+	syncAndValidateDaemonSets(t, manager, ds, podControl, 1, 0)
+}
+
+// DaemonSet should launch a pod on an untainted node when the pod has tolerations.
+func TestNodeDaemonLaunchesToleratePod(t *testing.T) {
+	manager, podControl, _ := newTestController()
+
+	node := newNode("untainted", nil)
+	manager.nodeStore.Add(node)
+
+	ds := newDaemonSet("tolerate")
+	setDaemonSetTeleration(ds)
+	manager.dsStore.Add(ds)
+
+	syncAndValidateDaemonSets(t, manager, ds, podControl, 1, 0)
+}
+
+func setNodeTaint(node *v1.Node) {
```

Comment:
nit: pass `noSchedule` as an argument to `setNodeTaint` and `setDaemonSetToleration` to make them reusable.